### PR TITLE
flow: Forbid object-as-map when constructing `Immutable.Map`s

### PR DESCRIPTION
**Pull request by chrisbobbe**
From discussion: https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.60Immutable.2EMap.28.7B.20.5BSTRING_CONST.5D.3A.20.E2.80.A6.20.7D.29.60.20busted.20by.20Flow.20bug/near/1481093

### CODE REVIEW

File: docs/style.md [hunk 1]
```diff
@@ -542,6 +542,25 @@ in Flow, or a design limitation of Flow, or an issue in the Flow types
...
+
+Immutable.Map({ [HOME_NARROW_STR]: [1, 2] }) // BAD -- don't do
+```
+
+This is essential in order to get effective type-checking of the value
+passed to the `Immutable.Map` function when a key is computed (like
+`HOME_NARROW_STR`) instead of literal (like 'foo'). In the Immutable type
+definition, the object-as-map branch involves an object type with an indexer
+property, and in general Flow seems to have trouble with those:
+  https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.60Immutable.2EMap.28.7B.20.5BSTRING_CONST.5D.3A.20.E2.80.A6.20.7D.29.60.20busted.20by.20Flow.20bug/near/1481088
```
Comment (gnprice) [inline docs/style.md:561]: This is Markdown, so let's make it a nice link:

```suggestion
property, and in general [Flow seems to have trouble with those][].

[Flow seems to have trouble with those]: https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.60Immutable.2EMap.28.7B.20.5BSTRING_CONST.5D.3A.20.E2.80.A6.20.7D.29.60.20busted.20by.20Flow.20bug/near/1481088
```

File: docs/style.md [hunk 2]
```diff
@@ -542,6 +542,25 @@ in Flow, or a design limitation of Flow, or an issue in the Flow types
 provided by Immutable.js, or some combination.)
 
 
+<div id="immutable-no-object-as-map" />
+
+**Don't construct an `Immutable.Map` with an object-as-map**: Whenever you
```
Comment (gnprice) [inline docs/style.md:547]: Should get an entry in the table of contents, following the pattern of the others.

(In particular, that provides the only convenient way to get these nice permalinks like `…#immutable-no-object-as-map` when viewing the document on GitHub.)

**Review by gnprice — COMMENTED** [review]
Thanks! Two style comments below on the style-guide entry.

I checked, and I believe this covers all the existing spots in the code where we had this pattern. The following command was handy:
```
$ git grep -P 'Immutable.Map(?!\(\[|\(\))'
```
(The incantation at the end filters out `Immutable.Map([` and `Immutable.Map()`.)

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

**Comment by gnprice** [comment]
Thanks! Looks good — merging.

**Pull request merged**